### PR TITLE
Fixes #13092 - Deploy correct bonding for reprovisioned host

### PR DESCRIPTION
**Pull request by xprazak2**

### CODE REVIEW

**Comment by xprazak2** [comment]
Integration tests failing

File: app/views/unattended/provisioning_templates/snippet/_kickstart_ifcfg_generic_interface.erb
Comment (ares) [inline app/views/unattended/provisioning_templates/snippet/_kickstart_ifcfg_generic_interface.erb]: please don't change the templates here, send a PR to [community-templates repo](https://github.com/theforeman/community-templates). Once it's merged there, we sync templates to Foreman code base from there. Change like this would be overidden on next sync.

File: app/views/unattended/provisioning_templates/snippet/_kickstart_networking_setup.erb
Comment (ares) [inline app/views/unattended/provisioning_templates/snippet/_kickstart_networking_setup.erb]: the same as above, should not be changed in this repo

File: app/views/unattended/provisioning_templates/snippet/_kickstart_ifcfg_generic_interface.erb
Comment (ares) [reply app/views/unattended/provisioning_templates/snippet/_kickstart_ifcfg_generic_interface.erb]: why is this change required? `@interface.vlanid` checks for tag and if it's not set, it loads the default value from subnet, this basically mean that only interfaces with explicit tag set would be considered vlans, while it's more common to have the tag set only for subnet which the vlan interface is attached to

File: app/views/unattended/provisioning_templates/snippet/_kickstart_networking_setup.erb
Comment (ares) [reply app/views/unattended/provisioning_templates/snippet/_kickstart_networking_setup.erb]: This would disable the real identifier detection for all interfaces, not just vlans attached to bonds. I suppose this should be only set if `@host.bond_interfaces.map(&:identifier).include?(interface.attached_to)` (untested)

File: app/models/host/base.rb
Comment (ares) [inline app/models/host/base.rb]: shouldn't this happen in `set_interface` just like `update_virtuals`? I suppose it should also happen only if identifier changed... maybe it should be combined with the `update_virtuals` method into single one

**Review by ares — CHANGES_REQUESTED** [review]

File: app/views/unattended/provisioning_templates/snippet/_kickstart_ifcfg_generic_interface.erb
Comment (xprazak2) [reply app/views/unattended/provisioning_templates/snippet/_kickstart_ifcfg_generic_interface.erb]: Because if `@subnet.nil? #=> true` and `@interface.tag.blank? #=> true`, then you get [undefined method 'vlanid' for nil](https://github.com/theforeman/foreman/blob/develop/app/models/nic/interface.rb#L30)

**Comment by xprazak2** [comment]
I moved the template changes into community-templates and updating the list of attached interfaces for a bond is now part of `set_interface` method.

File: app/models/host/base.rb
Comment (ShimShtein) [inline app/models/host/base.rb]: This forces a query to database for each bond, even if interfaces are already loaded. IMHO it would be better to use ruby's `select`:
``` ruby
interfaces.select { |iface| iface.mac == attributes['macaddress'] }
```

File: config/initializers/active_record_extensions.rb
Comment (tbrisker) [inline config/initializers/active_record_extensions.rb:11]: I think this could be a very dangerous addition - think about something like `host.interfaces.map(&:destroy)`

File: config/initializers/active_record_extensions.rb
Comment (tbrisker) [inline config/initializers/active_record_extensions.rb:15]: ditto

File: app/models/host/base.rb
Comment (tbrisker) [reply app/models/host/base.rb]: since we only need one `Enumerable#find` would be better then `Enumerable#select`.

**Review by tbrisker — DISMISSED** [review]

File: config/initializers/active_record_extensions.rb
Comment (ShimShtein) [reply config/initializers/active_record_extensions.rb:11]: We already have `#each`, I don't think `map` gives more power than each. It would be better to blacklist `#destroy`, since it's a dangerous action we don't want.

**Comment by xprazak2** [comment]
I replaces `find_by` with `detect`

**Comment by xprazak2** [comment]
katello failure unrelated

File: config/initializers/active_record_extensions.rb
Comment (tbrisker) [reply config/initializers/active_record_extensions.rb:11]: this has been addressed by latest safemode release.

File: app/models/host/base.rb
Comment (ares) [inline app/models/host/base.rb]: this should probably operate on iface passed as argument

File: app/models/host/base.rb
Comment (ares) [inline app/models/host/base.rb]: save should log errors if it fails

File: app/models/host/base.rb
Comment (ares) [inline app/models/host/base.rb]: this should live next to update_virtuals and only be triggered if identifier has changed (and similar conditions like update_virtual), the name could be update_bonds to make it clearer

**Review by ares — CHANGES_REQUESTED** [review]

**Comment by xprazak2** [comment]
I renamed `update_attached_interfaces` to `update_bonds`, passed it the interface that is already found, moved it further down to `update_virtuals` and added logging

**Comment by mmoll** [comment]
@xprazak2 could you rebase?

**Comment by xprazak2** [comment]
Rebased.

**Comment by mmoll** [comment]
[test katello]

**Comment by ares** [comment]
Some of my comments made this not working.

The update_bond needs to be executed before the interface identifier is changed, otherwise bond won't find `children_mac_addresses` because of attached devices search based on current interface identifier. So please, move it back under the comment you forgot to move with the update_bonds line, you had it right before my comments :-) at the same time, we can't do iface.identifier_changed?, at this point it was not changed yet. We can replace it with iface.identifier != name.

The error that was there is though the logger line which uses variables that are not available in that scope.

Since I know how much effort it involves to get the setup, I've sent you a PR with changes that made it working for me. This one is good to go when you merge.

@tbrisker any comments from your side? I'd like to merge when my comment above is resolved.

**Comment by ares** [comment]
here's the fixing PR for the reference https://github.com/xprazak2/foreman/pull/2

**Review by ares — APPROVED** [review]

**Comment by mmoll** [comment]
this needs a rebase for the audited pin

**Review by ares — CHANGES_REQUESTED** [review]
One quick rebase to make tests green please :-)

**Comment by xprazak2** [comment]
Rebased.

**Review by ares — APPROVED** [review]

**Comment by ares** [comment]
Thanks @xprazak2 and sorry for such long review time... merging!

**Pull request merged**
